src/components/Collection/CollectionView.tsx

```diff
@@ -0,0 +1,172 @@
+
+  const fetchPage = useMemo(() => debounce((page: number) => {
+    fetchGroups({ page, pageSize, filterId: filterId ?? '0' }).then((result) => {
+      if (!result.data) return;
+
+      const ids = result.data.pages[page].map(group => group.IDs.ID);
+      fetchGroupExtras({
+        GroupIDs: ids,
+        TagFilter: 128,
+        TagLimit: 20,
+        OrderByName: true,
+      }).then().catch(error => console.error(error));
+    }).catch(error => console.error(error)).finally(() => setFetchingPage(false));
+  }, 200), [filterId]);
+
+  useEffect(() => {
+    fetchPage.cancel();
+    setFetchingPage(false);
+
+    fetchPage(1);
+
+    return () => fetchPage.cancel();
+  }, [filterId]);
+
+  const { scrollRef } = useOutletContext<{ scrollRef: React.RefObject<HTMLDivElement> }>();
+
+  const [gridContainerRef, gridContainerBounds] = useMeasure();
+
+  const itemsPerRow = Math.max(1, Math.floor((gridContainerBounds.width + itemGap) / itemWidth));
+  const count = useMemo(() => Math.ceil(groupTotal / itemsPerRow), [groupTotal, itemsPerRow]);
+
+  const virtualizer = useVirtualizer({
+    count,
+    getScrollElement: () => scrollRef.current,
+    estimateSize: () => itemHeight,
+    overscan: 2,
+  });
+
+  if (groupTotal === 0) return (
+    <div className="flex grow items-center font-semibold justify-center">
+      {fetchingPage ? (
+        <Icon path={mdiLoading} size={3} className="text-highlight-1" spin />
+      ) : 'No series/groups available!'}
+    </div>
+  );
+
+  return (
+    <div className={cx(
+      'flex grow rounded-md',
+      mode === 'grid' && 'px-6 py-8 bg-background-alt border-background-border border',
+    )}>
+      <div className="w-full relative" style={{ height: virtualizer.getTotalSize() }} ref={gridContainerRef}>
+        {/* Each row is considered a virtual item here instead of each group */}
+        {virtualizer.getVirtualItems().map((virtualRow) => {
+          const index = virtualRow.index;
+
+          const items: React.ReactNode[] = [];
+          // index of the first group in the current row
+          // eg., row 2 (index = 1) at 1080p (itemsPerRow = 8)
+          // will have fromIndex as 8, meaning first group in row 2 is 8th in the group list
+          const fromIndex = index * itemsPerRow;
+          // index of the last group in the current row + 1
+          // same eg. as above, this will 16
+          const toIndex = fromIndex + itemsPerRow;
+
+          // Here, i will be the actual index of the group in group list
+          for (let i = fromIndex; i < toIndex; i++) {
+            const neededPage = Math.ceil((i + 1) / pageSize);
+            const relativeIndex = i % pageSize;
+            const groupList = groupPages[neededPage];
+
+            if (groupList === undefined && !fetchingPage) {
+              setFetchingPage(true);
+              fetchPage(neededPage);
+            }
+
+            const item = groupList !== undefined ? groupList[relativeIndex] : undefined;
```

Comment:
If you are on last/not full page, doesn't it mean you will start getting empty items here?

Reply:
`groupList[relativeIndex]` will return `undefined` and then in the next line `isPlaceholder` will become true, which is needed to render empty divs.
Without the placeholder divs, the items in the last row will be center aligned as each row has `justify-center` to not have more empty on the right than the left